### PR TITLE
chore(tools): Cache rework for `.circleci` jobs (part2)

### PR DESCRIPTION
**Pull request by zak-pawel**
**Improvement/fixes**

- For `windows`:
  - [windows-server-2022-gui](https://circleci.com/developer/machine/image/windows-server-2022-gui#how-this-image-works) contains preinstalled `go`, so we need to be sure that during `go` upgrade there will not be any clashes or leftovers (like we could observe [here](https://app.circleci.com/pipelines/github/influxdata/telegraf/17877/workflows/f63d747f-0715-4655-8ee8-a2ed73b8f529/jobs/276310)) so `/c/Program Files/Go` will be removed just before `restore_cache`. These clashes may also appear during `go` upgrade from previously cached version to newer version.
  - `MinGW` will also be cached and `/c/ProgramData/chocolatey/lib/mingw` will be removed just before `restore_cache` (for the same reason as for `go` mentioned above). `$PATH` manipulations were also needed (to check for original `gcc` inside `/c/ProgramData/chocolatey/lib/mingw/tools/install/mingw64/bin` and not for shim version in `/c/ProgramData/chocolatey/bin`) when installing `MinGW` and running tests with `-race`.
- For `darwin`:
  - Remove `/usr/local/Cellar/go` and create it with proper ownership just before `restore_cache`. It was done to fix a lot of `permission denied` errors which occurred [during cache restore](https://app.circleci.com/pipelines/github/influxdata/telegraf/18583/workflows/b2d1d451-bcb0-42cb-a01c-b423176af4f8/jobs/285639/parallel-runs/0/steps/0-104). Introduced during [update from `macos.x86.medium.gen2` to `macos.m1.medium.gen1`](https://github.com/influxdata/telegraf/pull/13932).
- For all **OS**es:
  - Names of cache keys unified to: 
     - `go-bins-<< parameters.cache_version >>-<< parameters.os >>-<< parameters.arch >>-go<< parameters.goversion >>-{{ checksum "go.sum" }}`
     - `go-caches-<< parameters.cache_version >>-<< parameters.os >>-<< parameters.arch >>-go<< parameters.goversion >>-{{ checksum "go.sum" }}`
   - Printing `$PATH`, binary (`go`/`gcc`) location and version added.

### CODE REVIEW

File: scripts/installmingw_windows.sh [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+MINGW_VERSION="12.2.0.03042023"
+GCC_VERSION="12.2.0"
+
+setup_mingw () {
+    choco upgrade mingw --allow-downgrade --force --version=${MINGW_VERSION}
+}
+
+export PATH="/c/ProgramData/chocolatey/lib/mingw/tools/install/mingw64/bin:$PATH"
+echo "$PATH"
```
Comment (powersj) [inline scripts/installmingw_windows.sh:13]: This is printed here and below, are both necessary?

**Review by powersj — COMMENTED** [review]
Thanks for tracking this all done. I have one extra print/echo below, otherwise, happy to see

File: scripts/installmingw_windows.sh (hunk 1 above)
Comment (zak-pawel) [reply scripts/installmingw_windows.sh:13]: @powersj Just to see what `chocolatey` added to the `$PATH`

**Review by srebhan — APPROVED** [review]
Nice job! Thanks for tackling this issue @zak-pawel!

**Review by powersj — APPROVED** [review]

**Pull request merged**